File: app/src/main/java/com/orionedwards/mtglifecounter/DataStore.java
package com.orionedwards.mtglifecounter;

import android.content.Context;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.NoSuchElementException;
import java.util.Scanner;

public class DataStore {
    public static JSONObject getWithKey(Context context, String key) throws DataStoreException {
        FileInputStream inputStream = null;
        try {
            inputStream = context.openFileInput(key + ".json");
            Scanner scanner = new Scanner(inputStream).useDelimiter("\\Z");
            String data = scanner.next();

            return new JSONObject(data);

        } catch (FileNotFoundException | JSONException | NoSuchElementException e) {
            throw new DataStoreException(e);
        } finally {
            try {
                if (inputStream != null)
                    inputStream.close();
            } catch (IOException unused) {} // what on earth can we do with an IOException upon file close
        }
    }

    public static void setWithKey(Context context, String key, JSONObject value) throws DataStoreException {
        FileOutputStream outputStream = null;
        OutputStreamWriter writer = null;
        try {
            outputStream = context.openFileOutput(key + ".json", Context.MODE_PRIVATE);
            writer = new OutputStreamWriter(outputStream);
            writer.write(value.toString());
        } catch (IOException e) {
            throw new DataStoreException(e);
        } finally {
            try {
                if (writer != null) {
                    writer.close();
                }
                if (outputStream != null) {
                    outputStream.close();
                }
            } catch (IOException unused){ }
        }
    }
}

class DataStoreException extends Exception {
    public DataStoreException(Exception inner) {
        super("data store exception", inner);
    }
}

File: app/src/main/java/com/orionedwards/mtglifecounter/DataStore.kt
package com.orionedwards.mtglifecounter

import android.content.Context

import org.json.JSONException
import org.json.JSONObject

import java.io.BufferedWriter
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStreamWriter
import java.util.NoSuchElementException
import java.util.Scanner

object DataStore {
    @Throws(DataStoreException::class)
    fun getWithKey(context: Context, key: String): JSONObject {
        var inputStream: FileInputStream? = null
        try {
            inputStream = context.openFileInput("$key.json")
            val scanner = Scanner(inputStream).useDelimiter("\\Z")
            val data = scanner.next()

            return JSONObject(data)

        } catch (e: FileNotFoundException) {
            throw DataStoreException(e)
        } catch (e: JSONException) {
            throw DataStoreException(e)
        } catch (e: NoSuchElementException) {
            throw DataStoreException(e)
        } finally {
            try {
                inputStream?.close()
            } catch (unused: IOException) {
            }
            // what on earth can we do with an IOException upon file close
        }
    }

    @Throws(DataStoreException::class)
    fun setWithKey(context: Context, key: String, value: JSONObject) {
        var outputStream: FileOutputStream? = null
        var writer: OutputStreamWriter? = null
        try {
            outputStream = context.openFileOutput("$key.json", Context.MODE_PRIVATE)
            writer = OutputStreamWriter(outputStream)
            writer.write(value.toString())
        } catch (e: IOException) {
            throw DataStoreException(e)
        } finally {
            try {
                writer?.close()
                outputStream?.close()
            } catch (unused: IOException) {
            }

        }
    }
}

internal class DataStoreException(inner: Exception) : Exception("data store exception", inner)
